x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java

```diff
@@ -53,22 +77,84 @@ public boolean foldable() {
     }
 
     @Override
-    public Boolean fold() {
-        if (Expressions.isNull(value()) || list().stream().allMatch(Expressions::isNull)) {
+    public Object fold() {
+        if (Expressions.isNull(value()) || list().size() == 1 && Expressions.isNull(list().get(0))) {
             return null;
         }
-        // QL's `In` fold() doesn't handle BytesRef and can't know if this is Keyword/Text, Version or IP anyway.
-        // `In` allows comparisons of same type only (safe for numerics), so it's safe to apply InProcessor directly with no implicit
-        // (non-numerical) conversions.
-        return InProcessor.apply(value().fold(), list().stream().map(Expression::fold).toList());
+        return EvaluatorMapper.super.fold();
+    }
+
+    @Override
+    public EvalOperator.ExpressionEvaluator.Factory toEvaluator(
+        Function<Expression, EvalOperator.ExpressionEvaluator.Factory> toEvaluator
+    ) {
+        var commonType = commonType();
+        EvalOperator.ExpressionEvaluator.Factory lhs;
+        EvalOperator.ExpressionEvaluator.Factory[] factories;
+        if (commonType.isNumeric()) {
+            lhs = Cast.cast(source(), value().dataType(), commonType, toEvaluator.apply(value()));
+            factories = list().stream()
+                .map(e -> Cast.cast(source(), e.dataType(), commonType, toEvaluator.apply(e)))
+                .toArray(EvalOperator.ExpressionEvaluator.Factory[]::new);
+        } else {
+            lhs = toEvaluator.apply(value());
+            factories = list().stream().map(e -> toEvaluator.apply(e)).toArray(EvalOperator.ExpressionEvaluator.Factory[]::new);
+        }
+
+        if (commonType == BOOLEAN) {
+            return new InBooleanEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == DOUBLE) {
+            return new InDoubleEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == INTEGER) {
+            return new InIntEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == LONG || commonType == DATETIME || commonType == UNSIGNED_LONG) {
+            return new InLongEvaluator.Factory(source(), lhs, factories);
+        }
+        if (commonType == KEYWORD
+            || commonType == TEXT
+            || commonType == IP
+            || commonType == VERSION
+            || commonType == UNSUPPORTED
```

Comment:
I'm a little confused as to what would cause `commonType` to be `UNSUPPORTED` (honestly, I'm a little confused as to what the `UNSUPPORTED` type is meant to do in general).  Could you add a comment here to clarify, please?

Reply:
> I'm a little confused as to what would cause `commonType` to be `UNSUPPORTED` (honestly, I'm a little confused as to what the `UNSUPPORTED` type is meant to do in general). Could you add a comment here to clarify, please?

That's a good question! I think I copied it from Greatest :-), I need to test more to understand the difference between the commonType UNSUPPERED and NULL, perhaps we should return a CONSTANT_NULL_FACTORY for both of them, I'll test it more.